File: pages/refactorings/replace-inheritance-with-natural-transformation/02-dictionary-set/Before.hs
-- Two phone books built by inserting entries in order (later
-- numbers win), then the entries both books agree on. A Dict is-a
-- Set of pairs, so one-value-per-key is re-imposed at every put.
module Before where

import Data.List (foldl')
import Data.Set (Set)
import qualified Data.Set as Set

type Dict k v = Set (k, v)

-- Ord v: the set orders values, which a map would never need.
put :: (Ord k, Ord v) => k -> v -> Dict k v -> Dict k v
put k v d = Set.insert (k, v) (Set.filter ((/= k) . fst) d)

fill :: [(String, Int)] -> Dict String Int
fill = foldl' (\d (k, v) -> put k v d) Set.empty

common :: [(String, Int)] -> [(String, Int)] -> Set (String, Int)
common xs ys = Set.intersection (fill xs) (fill ys)

File: pages/refactorings/replace-inheritance-with-natural-transformation/02-dictionary-set/Before.scala
// Two phone books built by inserting entries in order (later
// numbers win), then the entries both books agree on. A Dict is-a
// Set of pairs, so one-value-per-key is re-imposed at every put.
object Before:
  type Dict[K, V] = Set[(K, V)]

  def put[K, V](d: Dict[K, V], k: K, v: V): Dict[K, V] =
    d.filterNot(_._1 == k) + ((k, v))

  def fill(es: List[(String, Int)]): Dict[String, Int] =
    es.foldLeft(Set.empty[(String, Int)]) { (d, e) =>
      put(d, e._1, e._2)
    }

  def common(xs: List[(String, Int)],
             ys: List[(String, Int)]): Set[(String, Int)] =
    fill(xs) & fill(ys)
